torchvision/datasets/food101.py

```diff
@@ -0,0 +1,106 @@
+import json
+from pathlib import Path
+from typing import Any, Tuple, Callable, Optional
+
+import PIL.Image
+
+from .utils import verify_str_arg, download_and_extract_archive
+from .vision import VisionDataset
+
+
+class Food101(VisionDataset):
+    """`The Food-101 Data Set <https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101/>`_.
+
+    The Food-101 is a challenging data set of 101 food categories, with 101'000 images.
+    For each class, 250 manually reviewed test images are provided as well as 750 training images.
+    On purpose, the training images were not cleaned, and thus still contain some amount of noise.
+    This comes mostly in the form of intense colors and sometimes wrong labels. All images were
+    rescaled to have a maximum side length of 512 pixels.
+
+
+    Args:
+        root (string): Root directory of the dataset.
+        split (string, optional): The dataset split, supports ``"train"`` (default) and ``"test"``.
+
+        transform (callable, optional): A function/transform that  takes in an PIL image and returns a transformed
+            version. E.g, ``transforms.RandomCrop``.
+        target_transform (callable, optional): A function/transform that takes in the target and transforms it.
+    """
+
+    _URL = "http://data.vision.ee.ethz.ch/cvl/food-101.tar.gz"
+    _MD5 = "85eeb15f3717b99a5da872d97d918f87"
+
+    def __init__(
+        self,
+        root: str,
+        split: str = "train",
+        download: bool = True,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(root, transform=transform, target_transform=target_transform)
+        self._split = verify_str_arg(split, "split", ("train", "test"))
+        self._root_path = Path(self.root)
+
+        if download:
+            self._download()
+
+        if not self._check_exists():
+            raise RuntimeError("Dataset not found. You can use download=True to download it")
+
+        self._labels = []
+        self._image_files = []
+        with open(self._meta_folder / f"{split}.json", "r") as f:
+            metadata = json.loads(f.read())
+
+        self.classes = sorted(set(metadata.keys()))
+        self.class_to_idx = dict(zip(self.classes, range(len(self.classes))))
+
+        for class_label, im_paths in metadata.items():
+            self._labels += [self.class_to_idx[class_label]] * len(im_paths)
+            self._image_files += im_paths
```

Comment:
This is currently not working. The dataset does not store the actual file path, but rather its ID. Thus, we need to do three things to turn this into actual paths:

1. Add the `.jpg` suffix
2. Convert the `/` in the ID into the platform specific path component separator
3. Prefix `self._images_folder`

```suggestion
        for class_label, im_ids in metadata.items():
            self._labels += [self.class_to_idx[class_label]] * len(im_ids)
            self._image_files += [self._images_folder.joinpath(*f"{im_id}.jpg".split(os.sep)) for im_id in im_ids]
```

The tests currently pass, because you put the actual path in the meta data rather than the IDs.

Reply:
After looking closely while checking why the tests are failing I don't think we need to add the extension as it is already in the im_id (relative image path). Please look at my new commit and let me know if it makes sense for you

Reply:
> as it is already in the im_id

That is not true:

```python
import json

with open(".../food-101/meta/train.json") as file:
    meta = json.load(file)

print(meta[next(iter(meta))][0])
```

```
churros/1004234
```

Reply:
sorry, I was convinced that it did, but you are 100% right